Try sibling inserter tweaks

jasmussen opened the pull request:
This PR is based on feedback in #7168. It does this:

- It doesn't show the sibling inserter when hovering before the selected block.
- It shows the separator line when hovering the plus, never before.

This PR is a "try", and if we like it we'd want to make a few further improvements. Like we might actually want to allow showing the sibling inserter _before_ the selected block, if the user has chosen to dock the block toolbar to the top.

GIF:

![tweaks](https://user-images.githubusercontent.com/1204802/41146046-599c53dc-6b02-11e8-932e-92af343ab1ed.gif)

We might also want to look at improving it further by fading out hover outlines if you're hovering the sibling inserter. We'd need someone like @aduth or @youknowriad's help here.

CC: @folletto @SuperGeniusZeb @hedgefield @chrisvanpatten

Commit:
Try sibling inserter tweaks

This PR is based on feedback in #7168. It does this:

- It doesn't show the sibling inserter when hovering before the selected block.
- It shows the separator line when hovering the plus, never before.

This PR is a "try", and if we like it we'd want to make a few further improvements. Like we might actually want to allow showing the sibling inserter _before_ the selected block, if the user has chosen to dock the block toolbar to the top.

We might also want to look at improving it further by fading out hover outlines if you're hovering the sibling inserter. We'd need someone like @aduth or @youknowriad's help here.

chrisvanpatten commented:
I really like this! Definitely an improvement over what's in place now!

jasmussen commented:
It's interesting, I thought it would be disruptive to not be able to invoke the sibling inserter above the _selected block_, but it feels like a non issue in this branch.

chrisvanpatten commented:
One potential refinement: maybe the inserter should be visible above if the toolbar is pinned?

jasmussen commented:
Yep, definitely. It could also be cool to explore, potentially separately if need be, fading out the block outlines entirely if you're hovering the sibling inserter. Right now you can end in a situation where you're hovering both the sibling inserter, _and_ one block.

chrisvanpatten commented:
> It could also be cool to explore, potentially separately if need be, fading out the block outlines entirely if you're hovering the sibling inserter.

That could also help solve the overlap issue right now with narrow blocks (in columns, for instance):

![_untitled____add_new_page_ _mindful_ _wordpress](https://user-images.githubusercontent.com/1231306/41160207-ec8447b4-6afc-11e8-8f3d-b19945dc3677.png)

Also a "maybe another issue?" thought: should the inserter line go to the edges of the visible border?

SuperGeniusZeb commented:
This feels really nice! This is definitely an improvement over `master`, in my opinion.

One bug I noticed is that when you have selected a nested block, the sibling inserter does not appear for any of the other blocks in the same `<InnerBlocks/>`. You also cannot use the sibling inserter at the top of the parent block.

>Like we might actually want to allow showing the sibling inserter before the selected block, if the user has chosen to dock the block toolbar to the top.

Definitely. There is no reason not to show it when the block toolbar is docked, since there is nothing to conflict with the inserter.

>It's interesting, I thought it would be disruptive to not be able to invoke the sibling inserter above the selected block, but it feels like a non issue in this branch.

Yeah, since the block toolbar is already taking up space at the top of the selected block, it feels natural that the sibling inserter is not usable there.

>It could also be cool to explore, potentially separately if need be, fading out the block outlines entirely if you're hovering the sibling inserter.

I think this is a good idea.

@chrisvanpatten
>Also a "maybe another issue?" thought: should the inserter line go to the edges of the visible border?

Right now the block borders do not always match the actual size of the block since they are adjusted to make parent blocks always selectable. Right now I feel like the inserter line should continue to be the width of the actual block content, but I do not have any strong opinions on it either way.

karmatosed commented:
I don't want to rock in and suggest something isn't a good option, but I think we have to consider discoverability here. I think there is a little danger of having created now something only those that know where to hover would understand. What can we do to ease that?

afercia commented:
Maybe I'm missing something, but seems to me there's no way to use the sibling inserter with the keyboard?

chrisvanpatten commented:
> I think there is a little danger of having created now something only those that know where to hover would understand. What can we do to ease that? I feel using this it's less discoverable for those that don't use Gutenberg, we have to consider how used to it we are.

I don't think this PR actually changes any of the hover areas — the inserter button still shows up if you hover in the same areas as before. Unless you mean having the line visible all the time (instead of just when the button itself is hovered, which this PR does do), as an additional indication of change?

jasmussen commented:
Andrea:

> Maybe I'm missing something, but seems to me there's no way to use the sibling inserter with the keyboard?

You're not missing something. That's a regression. I went a little fast there with my try branch. Will fix.

Tammie:

> I don't want to rock in and suggest something isn't a good option, but I think we have to consider discoverability here. I think there is a little danger of having created now something only those that know where to hover would understand. What can we do to ease that? I feel using this it's less discoverable for those that don't use Gutenberg, we have to consider how used to it we are.

I strongly agree, and that is the genesis of #7168. I feel like _this pull request_ makes things much better, and we should get it in as soon as we can, as it simply refines who the existing sibling inserter works in `master`, so that it doesn't cause this to happen:

https://user-images.githubusercontent.com/1204802/41030245-286f5794-697e-11e8-8c1c-5057ef8cd675.png

However I also do not think that this PR closes #71698. It mitigates aspects, but the original issue is still present, and I think we should keep the discussion going there. Already a ton of good ideas, and some low hanging fruit like [adding shortcuts to the block's More menu](https://github.com/WordPress/gutenberg/issues/7168#issuecomment-395674089).

What do you think about getting this PR into master, as an improvement to the breakage that's present, and then continue to look at enhancements beyond this?

jasmussen commented:
Pushed a fix ot enable tabbing again. It's solid in my testing, but please give it a run through so we can see if it works as expected. If it does, I think this branch is ready for final review!

mtias commented:
I like the iterations here, it's coming back full circle :) With the new block outlines I feel the line on hover is out of place. I think we should remove it.

It's also looking like this feature will always be a bit of a "cool when you discover it". And I think that is fine for this feature. We'd probably want to add a "insert block below / above" to the ellipsis menu for completion purposes as it's one of those features where multiple avenues can benefit people with different flows.

What @jasmussen proposed here:

https://user-images.githubusercontent.com/1204802/41144730-23c68fce-6afe-11e8-8ab3-5322724c86bc.png

Commit:
Remove the line that appeared on hover.

Since it only appears on hover, it doesn't help discovery, and although the point is to show where you will insert things, it adds clutter next to the hovered outlines.

jasmussen commented:
Gave this one a good rebase and squash, and I removed the line per feedback:

![sibling](https://user-images.githubusercontent.com/1204802/41347428-ec9774ea-6f09-11e8-800d-bd7ac4f8f098.gif)

Created #7297 to address the menu items separately.

mtias commented:
This seems much nicer to me.

SuperGeniusZeb commented:
Neat, for some reason it never occurred to me earlier, but the borders of the hovered block can double as the line for the the inserter, making the line on hover redundant. It makes the UI look a bit lighter while not really losing anything.

There is still a bug where you cannot use the sibling inserter inside nested blocks when you have selected the block with nesting or its children.
![image](https://user-images.githubusercontent.com/19592990/41358144-636e270c-6eed-11e8-8bb1-86d3603fa618.png)

Also, I think you should be able to use the sibling inserter after the last block in a post/`<InnerBlocks/>`. This could be addressed in another PR, though.
![image](https://user-images.githubusercontent.com/19592990/41358091-3ee727e4-6eed-11e8-9aa8-236f24aeca56.png)
As shown in the above image, you can use the inserter in-between blocks in a nested context, but as shown in the image below, you cannot use it after the last block.
![image](https://user-images.githubusercontent.com/19592990/41358049-263552f2-6eed-11e8-93c0-0f7a6aac344f.png)

Other than those two points, I think this is ready for merge.

Commit:
Fix sibling inserter in nested contexts.

jasmussen commented:
Great feedback as usual. You've been super helpful in the past few months, Zeb, you too Chris. Very much appreciated 🏅 

> There is still a bug where you cannot use the sibling inserter inside nested blocks when you have selected the block with nesting or its children.

Fixed.

> Also, I think you should be able to use the sibling inserter after the last block in a post/<InnerBlocks/>. This could be addressed in another PR, though.

The sibling inserter is actually a component that's part each block, and sits before. So no "last blocks" actually have a sibling inserter that sits below them. That can be fixed of course, it would be if #6834 were addressed, but it's not a simple fix. It might also be mitigated if we always show the trailing appender after the last block, even in nested contexts — right now it's only shown if the last block is not a text block. Can't recall where this is ticketed.

So would think we should look at this separately.

karmatosed reviewed (approved):
I still have discoverability concerns in general about adding in-between elements. That all said, let's get this in and see it as a step in iterating. Approving on that basis.

mtias commented:
When clicking the plus, we should not retain the outline of the next block:

![image](https://user-images.githubusercontent.com/548849/41601612-331b6790-73d9-11e8-8fcc-89fb0ebe3d4d.png)

Commit:
Make plus blue.

jasmussen commented:
I made the plus blue, it looks better, and it also distinguishes this button a little bit from other Add Block buttons, which is good since the behavior is to insert a paragraph placeholder.

I explored this a bit further, with a new icon:

<img width="739" alt="screen shot 2018-06-20 at 09 44 28" src="https://user-images.githubusercontent.com/1204802/41644633-e9d14834-746e-11e8-8ba2-d00516a5750c.png">

This icon contains a line separator that indicates what's about to happen, and mimics the enhancements we've suggested could be added to the Block Menu:

<img width="332" alt="screen shot 2018-06-20 at 09 46 44" src="https://user-images.githubusercontent.com/1204802/41644650-fcd92136-746e-11e8-9ef7-01d9d7d25ee6.png">

Thoughts?

youknowriad committed:
Hide the outline when we click the in between inserter

jasmussen commented:
Thank you Riad for the help! This is how it looks now:

![sibling](https://user-images.githubusercontent.com/1204802/41651176-3c1edf28-7480-11e8-92d2-446cd8a7d2b2.gif)

I feel like we need to get this into master asap, as it's a VAAAST improvement over what's there. Further enhancements can be made, of course, there are many ideas to take a stab at. But this gets us 90% there, I feel. Final review?

hedgefield commented:
Yes, this looks real good! The only thing I'm fuzzy on is the + button on the left after you insert an empty block and defocus, what's that for?

jasmussen commented:
> The only thing I'm fuzzy on is the + button on the left after you insert an empty block and defocus, what's that for?

This is in `master` too, that's the in-canvas block library you access from there. You see it on the first line as well.

hedgefield commented:
I see, to me that one is a little confusing - I just clicked a + to add a block, why is there now a + in a different place that does a different thing?

Conceptually, I get that it opens the block library beyond the three suggestions on the blank block. But then I would find it more logical to have it as the last option after the three suggestions. "Do you want these common ones, or something else?" But that might be a separate ticket to explore.

mtias reviewed (approved):
Let's get this in. Nice work!

jasmussen commented:
Travis is complaining. I can rebase if need be?

aduth commented:
The test failure was intermittent. Passes now on a restart. See also https://github.com/WordPress/gutenberg/issues/6956#issuecomment-398894943.

SuperGeniusZeb commented:
@jasmussen Nice improvements! :smile:

I am not sure how I feel about the modified icon with the lines on the side. I do not dislike it, but I am not really sure I like it either. It feels redundant to me since the block border kind of doubles as a line for the icon, but I do not feel strongly either way about it.

@hedgefield I agree; it seems weird to have the inserter icon on the left. In nested contexts where the block suggestion buttons are not shown, the inserter icon is shown by itself within the borders of the placeholder block and on the right, rather than on the left and outside of the borders of the placeholder block. I think this inconsistency is a little confusing.
